### ISSUE TITLE
RNN support for Tensorflow Lite Micro

### ISSUE THREAD

**Issue by padoremu** [tensorflow/tflite-micro#907]
**System information**
- TensorFlow version: v1.12.1-23779-g96c5c8a 2.2.0-dev20200202
- Are you willing to contribute it: No

**Describe the feature and the current behavior/state.**
Support of RNNs is currently missing in Tensorflow Lite Micro. I've been testing with an RNN with GRU cells. Simple code (from [here](https://github.com/tensorflow/tensorflow/issues/33416#issuecomment-585081574)):

```python
import tensorflow as tf

model = tf.keras.Sequential()

model.add(tf.keras.layers.Input(shape=(1, 1,)))

cell = tf.keras.layers.GRUCell(10)

model.add(tf.keras.layers.RNN(cell, unroll=True))

converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.experimental_new_converter = True

tflite_model = converter.convert()

# for testing if operations are implemented by Tensorflow Lite
interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()
interpreter.invoke()
```

The current situation is (missing ops):
- `experimental_new_converter=True, unroll=True`: builtin ops SHAPE, TRANSPOSE, FILL, SPLIT_V, SUB, TANH
- `experimental_new_converter=True, unroll=False`: 3 subgraphs (only 1 subgraph supported)
- `experimental_new_converter=False, unroll=True`: builtin ops SPLIT_V, SUB, TANH
- `experimental_new_converter=False, unroll=False`: custom ops TensorListFromTensor, TensorListReserve, TensorListStack, While

My questions are:

1. Are there any tangible plans to implement the missing ops for RNNs for Tensorflow Lite Micro and if so, which one of the four variants will be the way to go?
2. Is it currently possible to somehow make the Toco converter for the non-unrolled case (i.e. `experimental_new_converter=False, unroll=False`) to convert the RNN just as a single (unsupported) placeholder RNN op rather than splitting it up into the four (unsupported) operators TensorListFromTensor, TensorListReserve, TensorListStack, While?

Thank you.

**Will this change the current api? How?**
No

**Who will benefit with this feature?**
Everybody who needs RNNs with Tensorflow Lite Micro.

**Any Other info.**
None

**Comment by advaitjain** [comment]
TFLM now supports many of the features (multiple subgraphs, additional OPs) that were blockers when this issue was first created. Please feel free to add any remaining missing features to this issue.

**Comment by juliusga** [comment]
@advaitjain Still facing the same issue

**Comment by StuartIanNaylor** [comment]
If anyone would like to provide a lstm for the new ESP32-S3 as think xtendsa or is cadence with the LX series have kept that layer behind a paywall.
For me on micro LSTM or GRU was a cul-de-sac for KWS which have some great LSTM/GRU models fave being CRNN as its lite compared to others for very similar accuracy.

Or do I have it wrong and the code does port to ESP32 specifically S3 as that has the vector instructions to make it much more model capable?

**Comment by advaitjain** [comment]
This particular issue is talking about RNN with GRU cells. The unimplemented OPs and multiple subgraph support are now part of the TFLM tree. My preference would be to limit this issue for issues related to GRU models.

Please create a separate issue for LSTM support - we are working on that and you can likely expect something in the coming months.

**Comment by trylaarsdam** [comment]
> TFLM now supports many of the features (multiple subgraphs, additional OPs) that were blockers when this issue was first created. Please feel free to add any remaining missing features to this issue.

Looks like even though there is LSTM support now in TFLM, the `TensorListFromTensor`, `TensorListReserve`, and `TensorListStack` ops are still not supported.

**Comment by kslavka** [comment]
Is there a support or tf.keras.layers.GRU in TFlite Micro, the same way it was done for UNIDIRECTIONAL_SEQUENCE_LSTM in 2022. 
Specifically, how to get a single GRU layer in TFLite model instead of unrolled GRU with multiple FullyConected, Mul, SPLIT, and Subs. 
Attaching a [link ](https://medium.com/@christoph-siegl/tensorflow-lite-for-microcontrollers-adds-support-for-efficient-lstm-implementation-25a5f7baa4f6) to an article talking about LSMT support in TFLM.
